--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -4860,8 +4860,8 @@
 
 static void prep_pager_sub_vars(struct ast_channel *ast, struct ast_vm_user *vmu, char *cidnum, char *dur, char *date)
 {
-	struct ast_config *msg_cfg;
-	struct ast_flags config_flags = { CONFIG_FLAG_NOCACHE };
+	//struct ast_config *msg_cfg;
+	//struct ast_flags config_flags = { CONFIG_FLAG_NOCACHE };
 
 	/* Prepare variables for substitution in pager body */
 	pbx_builtin_setvar_helper(ast, "VM_DUR", dur);
@@ -4869,7 +4869,7 @@
 	pbx_builtin_setvar_helper(ast, "VM_CIDNUM", (!ast_strlen_zero(cidnum) ? cidnum : "anonymous"));
 	pbx_builtin_setvar_helper(ast, "VM_DATE", date);
 
-	ast_config_destroy(msg_cfg);
+	// ast_config_destroy(msg_cfg);
 }
 
 static void prep_email_sub_vars(struct ast_channel *ast, struct ast_vm_user *vmu, int msgnum, char *context, char *mailbox, const char *fromfolder, char *cidnum, char *cidname, char *dur, char *date, const char *category, const char *flag)
@@ -4918,7 +4918,7 @@
 	}
 
 	if ((origtime = ast_variable_retrieve(msg_cfg, "message", "origtime")) && sscanf(origtime, "%30d", &inttime) == 1) {
-		struct timeval tv = { inttime, };
+		struct timeval tv = { inttime, 0 };
 		struct ast_tm tm;
 		ast_localtime(&tv, &tm, NULL);
 		ast_strftime_locale(origdate, sizeof(origdate), emaildateformat, &tm, S_OR(vmu->locale, NULL));
@@ -5305,7 +5305,7 @@
 				/* You might be tempted to do origdate, except that a) it's in the wrong
 				 * format, and b) it's missing for IMAP recordings. */
 				if ((v = ast_variable_retrieve(msg_cfg, "message", "origtime")) && sscanf(v, "%30d", &inttime) == 1) {
-					struct timeval tv = { inttime, };
+					struct timeval tv = { inttime, 0 };
 					struct ast_tm tm;
 					ast_localtime(&tv, &tm, NULL);
 					ast_strftime_locale(origdate, sizeof(origdate), emaildateformat, &tm, S_OR(vmu->locale, NULL));
--- a/Makefile
+++ b/Makefile
@@ -17,7 +17,8 @@
 
 LIBS+=-lpcre
 CFLAGS+=-pipe -fPIC
-CFLAGS+=-Wall -Wextra -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations
+CFLAGS+=-Wall -Wextra -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations \
+		-Wno-sign-compare -Wno-unused-parameter -Wno-unused-variabl
 CFLAGS+=-D_REENTRANT -D_GNU_SOURCE -DODBC_STORAGE
 
 all: _all
