From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_rename_apps

---
 Makefile | 20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index 5b96cba..67ec08d 100644
--- a/Makefile
+++ b/Makefile
@@ -45,13 +45,19 @@ apps/app_playback.so: apps/app_playback.o
 clean:
 	rm -f apps/app_voicemail.o apps/app_voicemail.so \
 		  apps/app_playback.o apps/app_playback.so
+	rm -f apps/apps_ngcp*.so
 
 install: _all
 	$(INSTALL) -m 755 -d $(DESTDIR)$(MODULES_DIR)
-	$(INSTALL) -m 755 apps/app_voicemail.so $(DESTDIR)$(MODULES_DIR)
-	$(INSTALL) -m 755 apps/app_playback.so $(DESTDIR)$(MODULES_DIR)
-	@echo " +---- apps/app_voicemail Installation Complete ------+"
-	@echo " +                                                    +"
-	@echo " + apps/app_voicemail has successfully been installed +"
-	@echo " +----------------------------------------------------+"
-
+	mv apps/app_voicemail.so apps/app_ngcp_voicemail.so
+	$(INSTALL) -m 755 apps/app_ngcp_voicemail.so $(DESTDIR)$(MODULES_DIR)
+	@echo " +---- apps/app_ngcp_voicemail Installation Complete ------+"
+	@echo " +                                                         +"
+	@echo " + apps/app_ngcp_voicemail has successfully been installed +"
+	@echo " +---------------------------------------------------------+"
+	mv apps/app_playback.so apps/app_ngcp_playback.so
+	$(INSTALL) -m 755 apps/app_ngcp_playback.so $(DESTDIR)$(MODULES_DIR)
+	@echo " +---- apps/app_ngcp_playback Installation Complete -------+"
+	@echo " +                                                         +"
+	@echo " + apps/app_ngcp_playback has successfully been installed  +"
+	@echo " +---------------------------------------------------------+"
