--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -1316,7 +1316,7 @@ static int findreplen(const char *rep, i
 	while(*cp) {
 		if (*cp == '$' && isdigit(cp[1])) {
 			val = strtoul(&cp[1], &cp, 10);
-			if (val && val <= nmat + 1)
+			if (val && val <= nmat)
 				len += replen[val -1];
 			else
 				fprintf(stderr, "repl %d out of range\n", val);
@@ -1336,7 +1336,7 @@ static void doreplace(char *out, const c
 	while(*cp) {
 		if (*cp == '$' && isdigit(cp[1])) {
 			val = strtoul(&cp[1], &cp, 10);
-			if (val && val <= nmat + 1) {
+			if (val && val <= nmat) {
 				strncpy(out, repstr[val - 1], replen[val - 1]);
 				out += replen[val -1];
 			}
@@ -1383,7 +1383,7 @@ static char *pcre_subst(const pcre *ppat
 	int nmat;
 	int ovec[MAXCAPTURE * 3];
 	nmat = pcre_exec(ppat, extra, str, len, offset, options,
-		ovec, sizeof(ovec));
+		ovec, sizeof(ovec) / sizeof(*ovec));
 	if (nmat <= 0)
 		return NULL;
 	return(edit(str, len, rep, nmat, ovec));
