From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_fix_prev_message

---
 apps/app_voicemail.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index beda819..8f4f3af 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -11426,6 +11426,7 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 #endif
 
 	/* Add the vm_state to the active list and keep it active */
+	vms.curmsg = -1;
 	vms.lastmsg = -1;
 	vms.deleted = NULL;
 	vms.heard = NULL;
