--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -1799,7 +1799,7 @@
  *
  * \return The ast_vm_user structure for the user that was found.
  */
-static struct ast_vm_user *find_user_realtime(struct ast_vm_user *ivm, const char *context, const char *mailbox)
+static struct ast_vm_user *find_user_realtime(struct ast_vm_user *ivm, const char *context, const char *mailbox, int by_uuid)
 {
 	struct ast_variable *var;
 	struct ast_vm_user *retval;
@@ -1816,9 +1816,15 @@
 			ast_copy_string(retval->mailbox, mailbox, sizeof(retval->mailbox));
 		}
 		if (!context && ast_test_flag((&globalflags), VM_SEARCH)) {
-			var = ast_load_realtime("voicemail", "mailbox", mailbox, SENTINEL);
+			if (by_uuid)
+				var = ast_load_realtime("voicemail", "customer_id", mailbox, SENTINEL);
+			else
+				var = ast_load_realtime("voicemail", "mailbox", mailbox, SENTINEL);
 		} else {
-			var = ast_load_realtime("voicemail", "mailbox", mailbox, "context", context, SENTINEL);
+			if (by_uuid)
+				var = ast_load_realtime("customer_id", "customer_id", mailbox, "context", context, SENTINEL);
+			else
+				var = ast_load_realtime("voicemail", "mailbox", mailbox, "context", context, SENTINEL);
 		}
 		if (var) {
 			// ast_log (LOG_DEBUG,"call apply_options_full\n");
@@ -1880,7 +1886,7 @@
 #ifdef ODBC_STORAGE
 		vmu = find_user_realtime_by_alias(ivm, context, mailbox);
 #else
-		vmu = find_user_realtime(ivm, context, mailbox);
+		vmu = find_user_realtime(ivm, context, mailbox, 0);
 #endif
 	}
 	AST_LIST_UNLOCK(&users);
@@ -4151,6 +4157,7 @@
 		"where customer_id = ?";
 	struct generic_prepare_struct gps_uuid = { .sql = sql_uuid, .argc = 1, .argv = argv };
 	struct odbc_obj *obj = NULL;
+	int by_uuid = 1;
 	SQLHSTMT stmt = NULL;
 
 	obj = ast_odbc_request_obj(odbc_database, 0);
@@ -4165,6 +4172,8 @@
 		ast_log(LOG_NOTICE, "Failed to fetch mailbox for uuid '%s', falling back to alias search\n", alias);
 		SQLFreeHandle (SQL_HANDLE_STMT, stmt);
 
+		by_uuid = 0;
+
 		stmt = ast_odbc_prepare_and_execute(obj, generic_prepare, &gps);
 		if (!stmt) {
 			ast_log(LOG_WARNING, "SQL Execute error!\n[%s]\n\n", sql);
@@ -4191,7 +4200,11 @@
 	ast_odbc_release_obj(obj);
 
 	ast_log(LOG_NOTICE, "Found mailbox '%s' for alias '%s'\n", mailbox, alias);
-	return find_user_realtime(ivm, context, mailbox);
+
+	if (by_uuid)
+		find_user_realtime(ivm, context, alias, by_uuid);
+
+	return find_user_realtime(ivm, context, mailbox, by_uuid);
 }
 #endif
 
