From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 13:57:31 +0100
Subject: sipwise_vm_fix_odbc_retreive_file

---
 apps/app_voicemail.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index c7b55a5..6a7e6b9 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -4092,7 +4092,9 @@ static int retrieve_file(char *dir, int msgnum)
 				ast_debug(3, "Ignoring null category column in ODBC voicemail retrieve_file.\n");
 				continue;
 			} else if (!SQL_SUCCEEDED(res)) {
-				ast_log(AST_LOG_WARNING, "SQL Get Data error! coltitle=%s\n[%s]\n\n", coltitle, sql);
+				if (strcasecmp(coltitle, "call_id")) {
+					ast_log(AST_LOG_WARNING, "SQL Get Data error! coltitle=%s\n[%s]\n\n", coltitle, sql);
+				}
 				goto bail_with_handle;
 			}
 			if (strcasecmp(coltitle, "msgnum") && strcasecmp(coltitle, "dir")) {
@@ -8623,7 +8625,9 @@ static int play_message(struct ast_channel *chan, struct ast_vm_user *vmu, struc
 		}
 #endif
 		if ((res = wait_file(chan, vms, vms->fn)) < 0) {
+#ifndef ODBC_STORAGE
 			ast_log(AST_LOG_WARNING, "Playback of message %s failed\n", vms->fn);
+#endif
 			res = 0;
 		}
 		ast_test_suite_event_notify("USERPRESS", "Message: User pressed %c\r\nDTMF: %c",
