From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_define_module

---
 apps/app_playback.c  | 2 ++
 apps/app_voicemail.c | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/apps/app_playback.c b/apps/app_playback.c
index 0c9281a..f8d0abd 100644
--- a/apps/app_playback.c
+++ b/apps/app_playback.c
@@ -31,6 +31,8 @@
 
 #include "asterisk.h"
 
+#define AST_MODULE "app_playback"
+
 #include "asterisk/file.h"
 #include "asterisk/pbx.h"
 #include "asterisk/module.h"
diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index e554ee7..b1f5182 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -100,6 +100,8 @@
 #endif
 #endif
 
+#define AST_MODULE "voicemail"
+
 #include "asterisk/paths.h"	/* use ast_config_AST_SPOOL_DIR */
 #include <sys/time.h>
 #include <sys/stat.h>
