From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_define_module

---
 apps/app_playback.c  | 3 +++
 apps/app_voicemail.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/apps/app_playback.c b/apps/app_playback.c
index 6b93b03..613e8c8 100644
--- a/apps/app_playback.c
+++ b/apps/app_playback.c
@@ -29,6 +29,9 @@
 	<support_level>core</support_level>
  ***/
 
+#define AST_MODULE "app_playback"
+#define AST_MODULE_SELF_SYM __app_playback_sym
+
 #include "asterisk.h"
 
 #include "asterisk/file.h"
diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 714e4d9..e7d0be2 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -46,6 +46,9 @@
  * \verbinclude voicemail.conf.sample
  */
 
+#define AST_MODULE "app_voicemail"
+#define AST_MODULE_SELF_SYM __app_voicemail_sym
+
 #include "asterisk.h"
 
 #ifdef IMAP_STORAGE
