From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_externnotify_by_uuid

---
 apps/app_voicemail.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 209b825..ef55d17 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -6312,15 +6312,15 @@ static void run_externnotify(const char *context, const char *extension, const c
 			if (msg_time) {
 				ast_localtime(msg_time, &tm, timezonename);
 				ast_strftime(date, sizeof(date), "%Y-%m-%dT%H:%M:%S%z", &tm);
-				snprintf(arguments, sizeof(arguments), "%s %s %s %d %d %d %d %s %s %d &",
+				snprintf(arguments, sizeof(arguments), "%s %s %s %s %d %d %d %d %s %s %d &",
 						 externnotify, S_OR(context, "\"\""),
-						 number, newvoicemails,
+						 number, extension, newvoicemails,
 						 oldvoicemails, urgentvoicemails,
 						 msgnum, cidnum, date, duration);
 			} else { // original short notify
-				snprintf(arguments, sizeof(arguments), "%s %s %s %d %d %d &",
+				snprintf(arguments, sizeof(arguments), "%s %s %s %s %d %d %d &",
 						 externnotify, S_OR(context, "\"\""),
-						 number, newvoicemails,
+						 number, extension, newvoicemails,
 						 oldvoicemails, urgentvoicemails);
 			}
 			ast_debug(1, "Executing %s\n", arguments);
