From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_find_user_by_alias_prio_uuid

---
 apps/app_voicemail.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 26bd50e..d1b2f99 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -4138,26 +4138,26 @@ static struct ast_vm_user *find_user_realtime_by_alias(struct ast_vm_user *ivm,
 		ast_log(LOG_WARNING, "Failed to obtain database object for '%s'!\n", odbc_database);
 		return NULL;
 	}
-	stmt = ast_odbc_prepare_and_execute(obj, generic_prepare, &gps);
+	stmt = ast_odbc_prepare_and_execute(obj, generic_prepare, &gps_uuid);
 	if (!stmt) {
-		ast_log(LOG_WARNING, "SQL Execute error!\n[%s]\n\n", sql);
+		ast_log(LOG_WARNING, "SQL Execute error!\n[%s]\n\n", sql_uuid);
 		ast_odbc_release_obj(obj);
 		return NULL;
 	}
 	res = SQLFetch(stmt);
 	if ((res != SQL_SUCCESS) && (res != SQL_SUCCESS_WITH_INFO)) {
-		ast_log(LOG_NOTICE, "Failed to fetch mailbox for alias '%s', falling back to uuid search\n", alias);
+		ast_log(LOG_NOTICE, "Failed to fetch mailbox for uuid '%s', falling back to alias search\n", alias);
 		SQLFreeHandle (SQL_HANDLE_STMT, stmt);
 
-		stmt = ast_odbc_prepare_and_execute(obj, generic_prepare, &gps_uuid);
+		stmt = ast_odbc_prepare_and_execute(obj, generic_prepare, &gps);
 		if (!stmt) {
-			ast_log(LOG_WARNING, "SQL Execute error!\n[%s]\n\n", sql_uuid);
+			ast_log(LOG_WARNING, "SQL Execute error!\n[%s]\n\n", sql);
 			ast_odbc_release_obj(obj);
 			return NULL;
 		}
 		res = SQLFetch(stmt);
 		if ((res != SQL_SUCCESS) && (res != SQL_SUCCESS_WITH_INFO)) {
-			ast_log(LOG_NOTICE, "Failed to fetch mailbox for alias '%s' via uuid\n", alias);
+			ast_log(LOG_NOTICE, "Failed to fetch mailbox for alias '%s'\n", alias);
 			SQLFreeHandle (SQL_HANDLE_STMT, stmt);
 
 			ast_odbc_release_obj(obj);
