From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_play_prompt_on_change_to_empty_folder

---
 apps/app_voicemail.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index c0a9aee..c2d74df 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -11246,6 +11246,7 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 	signed char record_gain = 0;
 	int play_auto = 0;
 	int play_folder = 0;
+	int folder_change = 0;
 	int in_urgent = 0;
 	int nodelete = 0;
 #ifdef IMAP_STORAGE
@@ -11515,6 +11516,7 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 			cmd = vm_browse_messages(chan, &vms, vmu);
 			break;
 		case '2': /* Change folders */
+			folder_change = 1;
 			ast_test_suite_event_notify("CHANGEFOLDER", "Message: browsing to a different folder");
 			if (useadsi)
 				adsi_folders(chan, 0, "Change to folder...");
@@ -11848,6 +11850,16 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 			vms.starting = 1;
  			break;
 		default:	/* Nothing */
+			if (folder_change && vms.lastmsg == -1) {
+				if(!strcasecmp(ast_channel_language(chan), "ro")) {
+					ast_play_and_wait(chan, "vm-dir-empty");
+				} else {
+					res = ast_play_and_wait(chan, "vm-no");
+					if (!res)
+						res = ast_play_and_wait(chan, "vm-messages");
+				}
+			}
+			folder_change = 0;
 			ast_test_suite_event_notify("PLAYBACK", "Message: instructions");
 			cmd = vm_instructions(chan, vmu, &vms, 0, in_urgent, nodelete);
 			break;
