--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -6701,15 +6701,15 @@ static void run_externnotify(char *conte
 			if (msg_time) {
 				ast_localtime(msg_time, &tm, timezonename);
 				ast_strftime(date, sizeof(date), "%Y-%m-%dT%H:%M:%S%z", &tm);
-				snprintf(arguments, sizeof(arguments), "%s %s %s %d %d %d %d %s %s %d &",
+				snprintf(arguments, sizeof(arguments), "%s %s %s %s %d %d %d %d %s %s %d &",
 						 externnotify, S_OR(context, "\"\""),
-						 number, newvoicemails,
+						 number, extension, newvoicemails,
 						 oldvoicemails, urgentvoicemails,
 						 msgnum, cidnum, date, duration);
 			} else { // original short notify
-				snprintf(arguments, sizeof(arguments), "%s %s %s %d %d %d &",
+				snprintf(arguments, sizeof(arguments), "%s %s %s %s %d %d %d &",
 						 externnotify, S_OR(context, "\"\""),
-						 number, newvoicemails,
+						 number, extension, newvoicemails,
 						 oldvoicemails, urgentvoicemails);
 			}
 			ast_debug(1, "Executing %s\n", arguments);
