From: Sipwise Development Team <support@sipwise.com>
Date: Tue, 6 Dec 2022 14:51:13 +0100
Subject: sipwise_vm_fix_envelope_play

---
 apps/app_voicemail.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 371cdde..8a6d46c 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -12450,8 +12450,10 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 					}
 					cmd = 't';
 					break;
+				*/
 				case '3': // Envelope
 					if (vms.lastmsg > -1 && !vms.starting) {
+						ast_safe_sleep(chan,500);
 						cmd = advanced_options(chan, vmu, &vms, vms.curmsg, 3, record_gain);
 						if (cmd == ERROR_LOCK_PATH) {
 							res = cmd;
@@ -12462,6 +12464,7 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 					}
 					cmd = 't';
 					break;
+				/*
 				case '4': // Dialout
 					if (!ast_strlen_zero(vmu->dialout)) {
 						cmd = dialout(chan, vmu, NULL, vmu->dialout);
@@ -12494,6 +12497,7 @@ static int vm_execmain(struct ast_channel *chan, const char *data)
 					break;
 
 				default:
+					ast_safe_sleep(chan,500);
 					cmd = 0;
 					/*
 					if (!vms.starting) {
