From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 4 Nov 2024 15:37:29 +0100
Subject: sipwise_vm_play_anonymous

---
 apps/app_voicemail.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index ddf4e9c..47a7e55 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -8921,7 +8921,7 @@ static int play_message_callerid(struct ast_channel *chan, struct vm_state *vms,
 	/* Strip off caller ID number from name */
 	ast_debug(1, "VM-CID: composite caller ID received: %s, context: %s\n", cid, context);
 	ast_callerid_parse(cid, &name, &callerid);
-	if ((!ast_strlen_zero(callerid)) && strcmp(callerid, "Unknown")) {
+	if ((!ast_strlen_zero(callerid)) && strcmp(callerid, "Unknown") && !strstr(callerid, "anonymous")) {
 		/* Check for internal contexts and only */
 		/* say extension when the call didn't come from an internal context in the list */
 		for (i = 0 ; i < MAX_NUM_CID_CONTEXTS ; i++){
