From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_change_password_use_customer_id

---
 apps/app_voicemail.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 086732d..14dd6d6 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -1616,7 +1616,7 @@ static int change_password_realtime(struct ast_vm_user *vmu, const char *passwor
 	if (strlen(password) > 10) {
 		ast_realtime_require_field("voicemail", "password", RQ_CHAR, strlen(password), SENTINEL);
 	}
-	if (ast_update2_realtime("voicemail", "context", vmu->context, "mailbox", vmu->mailbox, SENTINEL, "password", password, SENTINEL) > 0) {
+	if (ast_update2_realtime("voicemail", "context", vmu->context, "customer_id", vmu->mailbox, SENTINEL, "password", password, SENTINEL) > 0) {
 		ast_test_suite_event_notify("PASSWORDCHANGED", "Message: realtime engine updated with new password\r\nPasswordSource: realtime");
 		ast_copy_string(vmu->password, password, sizeof(vmu->password));
 		res = 0;
