--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -1571,7 +1571,7 @@
 	if (strlen(password) > 10) {
 		ast_realtime_require_field("voicemail", "password", RQ_CHAR, strlen(password), SENTINEL);
 	}
-	if (ast_update2_realtime("voicemail", "context", vmu->context, "mailbox", vmu->mailbox, SENTINEL, "password", password, SENTINEL) > 0) {
+	if (ast_update2_realtime("voicemail", "context", vmu->context, "customer_id", vmu->mailbox, SENTINEL, "password", password, SENTINEL) > 0) {
 		ast_test_suite_event_notify("PASSWORDCHANGED", "Message: realtime engine updated with new password\r\nPasswordSource: realtime");
 		ast_copy_string(vmu->password, password, sizeof(vmu->password));
 		res = 0;
