---
 apps/app_playback.c  |    2 ++
 apps/app_voicemail.c |    2 ++
 2 files changed, 4 insertions(+)

--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -80,6 +80,8 @@
 </category>
 ***/
 
+#define AST_MODULE "voicemail"
+
 #include "asterisk.h"
 
 #ifdef IMAP_STORAGE
--- a/apps/app_playback.c
+++ b/apps/app_playback.c
@@ -29,6 +29,8 @@
 	<support_level>core</support_level>
  ***/
 
+#define AST_MODULE "app_playback"
+
 #include "asterisk.h"
 
 ASTERISK_FILE_VERSION(__FILE__, "$Revision$")
