From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_vm_hebrew_language_fixups

---
 apps/app_voicemail.c | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 97f1f54..3064599 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -8510,10 +8510,15 @@ static int play_message(struct ast_channel *chan, struct ast_vm_user *vmu, struc
 	vms->starting = 0;
 	make_file(vms->fn, sizeof(vms->fn), vms->curdir, vms->curmsg);
 	adsi_message(chan, vms);
-	if (!vms->curmsg) {
-		res = wait_file2(chan, vms, "vm-first");	/* "First" */
-	} else if (vms->curmsg == vms->lastmsg) {
-		res = wait_file2(chan, vms, "vm-last");		/* "last" */
+
+	if (!strncasecmp(ast_channel_language(chan), "he", 2)) {
+		/* Do not play "vm-first"/"vm-last" for Hebrew syntaxes */
+	} else {
+		if (!vms->curmsg) {
+			res = wait_file2(chan, vms, "vm-first");	/* "First" */
+		} else if (vms->curmsg == vms->lastmsg) {
+			res = wait_file2(chan, vms, "vm-last");		/* "last" */
+		}
 	}
 
 	snprintf(filename, sizeof(filename), "%s.txt", vms->fn);
