--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -80,6 +80,9 @@
 </category>
 ***/
 
+#define AST_MODULE "app_voicemail"
+#define AST_MODULE_SELF_SYM __app_voicemail_sym
+
 #include "asterisk.h"
 
 #ifdef IMAP_STORAGE
@@ -101,8 +104,6 @@
 #endif
 #endif
 
-#define AST_MODULE "voicemail"
-
 #include "asterisk/paths.h"	/* use ast_config_AST_SPOOL_DIR */
 #include <sys/time.h>
 #include <sys/stat.h>
@@ -11865,7 +11866,7 @@ static int vm_authenticate(struct ast_ch
 				ast_log(LOG_NOTICE, "Rewrote mailbox user input '%s' to %s\n",
 						mailbox, normalized_mailbox);
 				ast_copy_string(mailbox, normalized_mailbox, mailbox_size);
-				free(normalized_mailbox);
+				ast_free(normalized_mailbox);
 			}
 		}
 
@@ -12113,8 +12114,8 @@ static int vm_execmain(struct ast_channe
 	int box;
 	int useadsi = 0;
 	int skipuser = 0;
-	struct vm_state vms = {{0}};
-	struct ast_vm_user *vmu = NULL, vmus = {{0}};
+	struct vm_state vms;
+	struct ast_vm_user *vmu = NULL, vmus;
 	char *context = NULL;
 	int silentexit = 0;
 	struct ast_flags flags = { 0 };
@@ -13452,7 +13453,7 @@ static struct ast_custom_function vm_inf
 static int vmauthenticate(struct ast_channel *chan, const char *data)
 {
 	char *s, *user = NULL, *context = NULL, mailbox[AST_MAX_EXTENSION] = "";
-	struct ast_vm_user vmus = {{0}};
+	struct ast_vm_user vmus = {};
 	char *options = NULL;
 	int silent = 0, skipuser = 0;
 	int res = -1;
@@ -17409,5 +17410,7 @@ AST_MODULE_INFO(ASTERISK_GPL_KEY, AST_MO
 	.load = load_module,
 	.unload = unload_module,
 	.reload = reload,
+#if 0
 	.optional_modules = "res_adsi,res_smdi",
+#endif
 );
--- a/apps/app_playback.c
+++ b/apps/app_playback.c
@@ -29,9 +29,10 @@
 	<support_level>core</support_level>
  ***/
 
-#include "asterisk.h"
-
 #define AST_MODULE "app_playback"
+#define AST_MODULE_SELF_SYM __app_playback_sym
+
+#include "asterisk.h"
 
 #include "asterisk/file.h"
 #include "asterisk/pbx.h"
--- a/Makefile
+++ b/Makefile
@@ -18,7 +18,7 @@ DEBUG=-g
 LIBS+=-lpcre
 CFLAGS+=-pipe -fPIC
 CFLAGS+=-Wall -Wextra -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations \
-		-Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable
+		-Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-format-truncation
 CFLAGS+=-D_REENTRANT -D_GNU_SOURCE -DODBC_STORAGE
 
 all: _all
