From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 4 Nov 2024 15:37:29 +0100
Subject: sipwise_vm_fix_odbc_retreive_file

---
 apps/app_voicemail.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index f6432de..691b730 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -4028,7 +4028,7 @@ static void odbc_update_msg_id(char *dir, int msg_num, char *msg_id)
 	char *argv[] = { msg_id, dir, msg_num_str };
 	struct generic_prepare_struct gps = { .sql = sql, .argc = 3, .argv = argv };
 
-	obj = ast_odbc_request_obj(odbc_database, 0);
+    obj = ast_odbc_request_obj(odbc_database, 0);
 	if (!obj) {
 		ast_log(LOG_WARNING, "Unable to update message ID for message %d in %s\n", msg_num, dir);
 		return;
@@ -4199,6 +4199,9 @@ static int retrieve_file(char *dir, int msgnum)
 				continue;
 			} else if (!SQL_SUCCEEDED(res)) {
 				ast_log(AST_LOG_WARNING, "SQL Get Data error! coltitle=%s\n[%s]\n\n", coltitle, sql);
+				if (strcasecmp(coltitle, "call_id")) {
+					ast_log(AST_LOG_WARNING, "SQL Get Data error! coltitle=%s\n[%s]\n\n", coltitle, sql);
+				}
 				goto bail_with_handle;
 			}
 			if (strcasecmp(coltitle, "msgnum") && strcasecmp(coltitle, "dir")) {
@@ -9407,7 +9410,9 @@ static int play_message(struct ast_channel *chan, struct ast_vm_user *vmu, struc
 		}
 #endif
 		if ((res = wait_file(chan, vms, vms->fn)) < 0) {
+#ifndef ODBC_STORAGE
 			ast_log(AST_LOG_WARNING, "Playback of message %s failed\n", vms->fn);
+#endif
 			res = 0;
 		}
 		ast_test_suite_event_notify("USERPRESS", "Message: User pressed %c\r\nDTMF: %c",
