--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -144,6 +144,8 @@ ASTERISK_FILE_VERSION(__FILE__, "$Revisi
 #include "asterisk/threadstorage.h"
 #endif
 
+#define AST_MODULE "voicemail"
+
 /*** DOCUMENTATION
 	<application name="VoiceMail" language="en_US">
 		<synopsis>
--- a/apps/app_playback.c
+++ b/apps/app_playback.c
@@ -33,6 +33,8 @@
 
 ASTERISK_FILE_VERSION(__FILE__, "$Revision$")
 
+#define AST_MODULE "app_playback"
+
 #include "asterisk/file.h"
 #include "asterisk/pbx.h"
 #include "asterisk/module.h"
