--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -12450,8 +12450,10 @@ static int vm_execmain(struct ast_channe
 					}
 					cmd = 't';
 					break;
+				*/
 				case '3': // Envelope
 					if (vms.lastmsg > -1 && !vms.starting) {
+						ast_safe_sleep(chan,500);
 						cmd = advanced_options(chan, vmu, &vms, vms.curmsg, 3, record_gain);
 						if (cmd == ERROR_LOCK_PATH) {
 							res = cmd;
@@ -12462,6 +12464,7 @@ static int vm_execmain(struct ast_channe
 					}
 					cmd = 't';
 					break;
+				/*
 				case '4': // Dialout
 					if (!ast_strlen_zero(vmu->dialout)) {
 						cmd = dialout(chan, vmu, NULL, vmu->dialout);
@@ -12494,6 +12497,7 @@ static int vm_execmain(struct ast_channe
 					break;
 
 				default:
+					ast_safe_sleep(chan,500);
 					cmd = 0;
 					/*
 					if (!vms.starting) {
