From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 21 Feb 2022 14:17:23 +0100
Subject: sipwise_fix_chan_usage

---
 apps/app_voicemail.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index c72c9e5..8e6e343 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -15516,7 +15516,9 @@ AST_TEST_DEFINE(test_voicemail_msgcount)
 			break;
 		}
 		open_mailbox(&vms, vmu, folder2mbox[i]);
+		#ifdef IMAP_STORAGE
 		STORE(tmp[i].dir, testmailbox, testcontext, 0, chan, vmu, "gsm", 600, &vms, strcmp(folders[i], "Urgent") ? "" : "Urgent", NULL);
+		#endif
 
 		/* hasvm-old, hasvm-urgent, hasvm-new, ic-old, ic-urgent, ic-new, ic2-old, ic2-urgent, ic2-new, mc-old, mc-urgent, mc-new */
 		for (j = 0; j < 3; j++) {
