From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 4 Nov 2024 15:37:29 +0100
Subject: sipwise_vm_hebrew_language_fixups

---
 apps/app_voicemail.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index c825fee..c969cc2 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -9276,10 +9276,14 @@ static int play_message(struct ast_channel *chan, struct ast_vm_user *vmu, struc
 	vms->starting = 0;
 	make_file(vms->fn, sizeof(vms->fn), vms->curdir, vms->curmsg);
 	adsi_message(chan, vms);
-	if (!vms->curmsg) {
-		res = wait_file2(chan, vms, "vm-first");	/* "First" */
-	} else if (vms->curmsg == vms->lastmsg) {
-		res = wait_file2(chan, vms, "vm-last");		/* "last" */
+
+	/* NOT HEBREW syntax */
+	if (strncasecmp(ast_channel_language(chan), "he", 2)) {
+		if (!vms->curmsg) {
+			res = wait_file2(chan, vms, "vm-first");	/* "First" */
+		} else if (vms->curmsg == vms->lastmsg) {
+			res = wait_file2(chan, vms, "vm-last");		/* "last" */
+		}
 	}
 
 	snprintf(filename, sizeof(filename), "%s.txt", vms->fn);
